### PR TITLE
Fixed #30071 -- Fixed error message when no default db is provided.

### PR DESCRIPTION
**Pull request by benjyw**
In ConnectionHandler, the explicit check for the DEFAULT_DB_ALIAS
key in the databases dict came after an attempt to access that key.

If you have at least one non-default db, but no default db, no dummy
default will be created for you, and the DEFAULT_DB_ALIAS key will
not be present. Under these circumstances, calling the `databases`
property directly would give a KeyError instead of the intended
ImproperlyConfigured error.

Worse, if you tried to access a non-default db, ensure_defaults would
catch that KeyError but misinterpret it as being for the non-default
db name (which does in fact exist in the dict).

(See https://code.djangoproject.com/ticket/30071)

This change moves the explicit check before the attempt to use the key,
so that no KeyError is ever raised for DEFAULT_DB_ALIAS.  It adds
a test for this case, and also adds a test for the case where the
default db is explicitly set to an empty dict.

### CODE REVIEW

File: tests/db_utils/tests.py [hunk 1]
```diff
@@ -21,6 +21,27 @@ def test_connection_handler_no_databases(self):
...
+        """Empty default database setting defaults to the dummy backend."""
+        DATABASES = {DEFAULT_DB_ALIAS: {}}
+        conns = ConnectionHandler(DATABASES)
+        self.assertEqual(conns[DEFAULT_DB_ALIAS].settings_dict['ENGINE'], 'django.db.backends.dummy')
+        msg = (
+            'settings.DATABASES is improperly configured. Please supply the '
+            'ENGINE value. Check settings documentation for more details.'
+        )
+        with self.assertRaisesMessage(ImproperlyConfigured, msg):
+            conns[DEFAULT_DB_ALIAS].ensure_connection()
```
Comment (felixxm) [inline tests/db_utils/tests.py]: This test is not related with the fix, so it will be better to create a separate PR for increasing tests coverage. I would add `DATABASES = {DEFAULT_DB_ALIAS: {}}` as a next case to the `test_connection_handler_no_databases()` (probably renamed), e.g.
```python
def test_connection_handler_no_databases(self):
    for DATABASES in (
        {},  # Empty DATABASES setting.
        {DEFAULT_DB_ALIAS: {}},  # Empty default database setting.
    ):
         conns = ConnectionHandler(DATABASES)
         ...
```

Comment (benjyw) [reply tests/db_utils/tests.py]: Removed the unrelated test.
